src/main/java/teammates/ui/webapi/action/GetStudentProfileAction.java

```diff
@@ -16,15 +18,55 @@ protected AuthType getMinAuthLevel() {
 
     @Override
     public void checkSpecificAccessControl() {
-        if (!userInfo.isStudent) {
-            throw new UnauthorizedAccessException("Student privilege is required to access this resource.");
+
+        String entityType = getNonNullRequestParamValue(Const.ParamsNames.ENTITY_TYPE);
+        if (userInfo.isStudent && entityType.equals(Const.EntityType.STUDENT)) {
+            checkStudentAccess();
+        } else if (userInfo.isInstructor && entityType.equals(Const.EntityType.INSTRUCTOR)) {
+            checkInstructorAccess();
+        } else {
+            throw new UnauthorizedAccessException("Your account cannot access this student's profile.");
+        }
+    }
+
+    private void checkStudentAccess() {
+        String studentId = getNonNullRequestParamValue(Const.ParamsNames.STUDENT_ID);
+        if (studentId.equals(userInfo.id)) {
+            // Student wants to access his own profile, permission always granted.
+            return;
         }
+        String courseId = getNonNullRequestParamValue(Const.ParamsNames.COURSE_ID);
+
+        StudentAttributes studentUser = logic.getStudentForGoogleId(courseId, userInfo.id);
+        if (studentUser == null) {
+            throw new UnauthorizedAccessException("You are not the classmate of this student,"
+                    + " cannot access his profile");
+        }
+
+        StudentAttributes student = getNonNullStudentByGoogleId(studentId, courseId);
+        if (!studentUser.team.equals(student.team)) {
+            throw new UnauthorizedAccessException("You are not the teammate of this student,"
```

Comment:
Same change in output here. Additionally, I notice that the strings are very similar, perhaps they could be abstracted out to Const?